[FLINK-26107][runtime] AdaptiveScheduler: OperatorCoordinator should …

dmvk opened the pull request:
This fixes a regression introduced in https://github.com/apache/flink/commit/5180e83ec01236ecce47a56951daf2a34eb4a1f6, that makes several OC related tests fail with the AdaptiveScheduler (unfortunately with the CRON triggered CI).

The intuition behind the change is that the OC failure handling is tied with the global failure handler of the CreatingExecutionGraph state, but we need it to use a global failure handler of the State that is currently in effect (eg. Executing).

I've added new unit test to guard against future regression.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraph.java:
```diff
@@ -205,6 +218,14 @@ AssignmentResult tryToAssignSlots(
+         *
+         * @return The current state of the {@link AdaptiveScheduler}.
+         */
+        State getEffectiveState();
```
I did that in the first iteration, but it seemed confusing that you actually have both `handleGlobalFailure` and `context.handleGlobalFailure` methods available in the class and it's not clear what the difference is. This felt bit more explicit for the reader.

WDYT?

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:
```diff
@@ -149,9 +159,27 @@ public void testSuccessfulSlotAssignmentTransitionsToExecuting() throws Exceptio
                             executionGraph, new TestingVertexParallelism()));
+
+            final Throwable t = new RuntimeException("Test.");
+            errorHandlerRef.get().accept(t);
```
This should be a separate test case. It now contains more lines for the error handling although that's not it's primary purpose.

dmvk replied:
👍

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:
```diff
@@ -149,9 +159,27 @@ public void testSuccessfulSlotAssignmentTransitionsToExecuting() throws Exceptio
+
+            final Throwable t = new RuntimeException("Test.");
+            errorHandlerRef.get().accept(t);
+            assertEquals(
+                    "OperatorCoordinator should use failure handler of the Executing state.",
+                    t,
+                    handledFailureRef.get());
```
This overall seems unsafe. There's no requirement that the error handler for different states are actually different instances.
What you rather want to test is that a failure in the coordinator calls into the scheduler.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:
```diff
@@ -135,10 +139,16 @@ public void testNotPossibleSlotAssignmentTransitionsToWaitingForResources() thro
+            final AtomicReference<Consumer<Throwable>> errorHandlerRef = new AtomicReference<>();
+            new CreatingExecutionGraph(context, executionGraphWithVertexParallelismFuture, log) {
+                @Override
+                OperatorCoordinatorHandler createOperatorCoordinatorHandler(
```
I'd rather add a OperatorCoordinatorHandler factory argument to the CreatingExecutionGraph than resorting to this.

dmvk replied:
👍

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:
```diff
@@ -149,9 +159,27 @@ public void testSuccessfulSlotAssignmentTransitionsToExecuting() throws Exceptio
+
+                                @Override
+                                public void handleGlobalFailure(Throwable cause) {
+                                    handledFailureRef.set(cause);
```
This seems overly complicated; isn't it enough to know that it was indeed called?

dmvk commented:
Thanks for the review @zentol! I've tried to go bit more fancy with the PR, let me know what you think (it's basically a complete rework 😂 )

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:
```diff
@@ -155,6 +179,60 @@ public void testSuccessfulSlotAssignmentTransitionsToExecuting() throws Exceptio
+                    });
+
+            final AtomicBoolean contextGlobalFailureHandlerCalled = new AtomicBoolean(false);
+            context.setGlobalFailureHandler(
```
Couldn't we just check that the errorHandler passed to the factory == context?

dmvk replied:
I'd rather avoid that, because we can't do an assertion in factory as it's executed in the future callback that is wrapped in `FutureUtils.assertNoException`. I don't really want to go into changing any more logic because of that.

I can still make an assert on the reference, but that doesn't seem to make too much of a difference for readability.

WDYT?

dmvk commented:
CI has failed with https://issues.apache.org/jira/browse/FLINK-20928

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:
```diff
@@ -155,6 +179,60 @@ public void testSuccessfulSlotAssignmentTransitionsToExecuting() throws Exceptio
+                        // This just lets us transition to executing.
+                    });
+
+            final AtomicBoolean contextGlobalFailureHandlerCalled = new AtomicBoolean(false);
+            context.setGlobalFailureHandler(
+                    t -> {
+                        contextGlobalFailureHandlerCalled.set(true);
+                    });
+
+            executionGraphWithVertexParallelismFuture.complete(
+                    CreatingExecutionGraph.ExecutionGraphWithVertexParallelism.create(
+                            executionGraph, new TestingVertexParallelism()));
+
+            operatorCoordinatorGlobalFailureHandlerRef
+                    .get()
+                    .handleGlobalFailure(new RuntimeException("Test."));
+            assertTrue(contextGlobalFailureHandlerCalled.get());
```
This is what I meant in https://github.com/apache/flink/pull/18761#discussion_r806234796.

```suggestion
            executionGraphWithVertexParallelismFuture.complete(
                    CreatingExecutionGraph.ExecutionGraphWithVertexParallelism.create(
                            executionGraph, new TestingVertexParallelism()));

            assertThat(operatorCoordinatorGlobalFailureHandlerRef).isSameAs(context);
```

dmvk replied:
👍 will change that

dmvk commented:
@zentol I've addressed your comment. The last change is done in a separate commit as it also contains migration to junit5 so we don't introduce too much noise into the actual change.
